src/TorchSharp/DataLoader.cs

```diff
@@ -0,0 +1,114 @@
+                    /// <param name="batchSize"></param>
+                    /// <param name="shuffle"></param>
+                    /// <param name="device"></param>
+                    public DataLoader(Dataset dataset, int batchSize, bool shuffle = false, Device device = null)
+                    {
+                        this.dataset = dataset;
+                        this.batchSize = batchSize;
+                        this.shuffle = shuffle;
+                        this.device = device ?? CPU;
+                    }
+
+                    public IEnumerator<Dictionary<string, Tensor>> GetEnumerator() =>
+                        new DataLoaderEnumerator(dataset, batchSize, shuffle, device);
+
+                    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+                    public long Count => (dataset.Count - 1) / batchSize + 1;
+
+                    private class DataLoaderEnumerator : IEnumerator<Dictionary<string, Tensor>>
+                    {
+                        private Dataset dataset;
+                        private int batchSize;
+                        private Device device;
+                        private bool shuffle;
+                        private ShuffleGenerator shuffleGenerator;
+                        private int currentVal = 0;
+
+                        public DataLoaderEnumerator(Dataset dataset, int batchSize, bool shuffle, Device device)
+                        {
+                            this.dataset = dataset;
+                            this.batchSize = batchSize;
+                            this.device = device;
+                            this.shuffle = shuffle;
+                            reset();
+                        }
+
+                        private bool isFinished() =>
+                            shuffle ? !shuffleGenerator.hasNext() : currentVal >= dataset.Count;
+
+                        private int getNextValue() => shuffle ? shuffleGenerator.next() : currentVal++;
+
+                        private void reset()
+                        {
+                            shuffleGenerator = new ShuffleGenerator(dataset.Count);
+                            currentVal = 0;
+                        }
+
+                        public bool MoveNext()
+                        {
+                            if (isFinished()) return false;
+                            Current = dataset.GetTensor(getNextValue());
+                            var currentKeys = Current.Keys;
+                            foreach (var x in currentKeys)
+                                Current[x].unsqueeze_(0);
+                            Dictionary<string, Tensor> dic;
+                            for (var i = 1; i < batchSize; i++) {
+                                if (isFinished())
+                                    break;
+                                dic = dataset.GetTensor(getNextValue());
+                                foreach (var x in currentKeys)
```

Comment:
Wouldn't it be better to gather all the unsqueezed tensors in a list (or array) and then catenate them all together at the end. It would seem a lot more efficient, and will create fewer temporaries.